### PR TITLE
Add a new HashSet template

### PR DESCRIPTION
**Pull request by reduz**
* Intended to replace RBSet in most cases.
* Optimized for iteration speed

### CODE REVIEW

File: core/templates/hash_set.h [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+				distance = existing_probe_len;
+			}
+
+			pos = (pos + 1) % capacity;
+			distance++;
+		}
+	}
+
+	void _resize_and_rehash(uint32_t p_new_capacity_index) {
+		uint32_t old_capacity = hash_table_size_primes[capacity_index];
```
Comment (bruvzg) [inline core/templates/hash_set.h:141]:
```
./core/templates/hash_set.h:141:12: error: unused variable 'old_capacity' [-Werror,-Wunused-variable]
                uint32_t old_capacity = hash_table_size_primes[capacity_index];
```

File: tests/test_main.cpp [hunk 2]
```diff
@@ -60,6 +60,7 @@
 #include "tests/core/string/test_translation.h"
 #include "tests/core/templates/test_command_queue.h"
 #include "tests/core/templates/test_hash_map.h"
+#include "tests/core/templates/test_hash_set.h"
```
Comment (bruvzg) [inline tests/test_main.cpp:63]: Missing file.

**Comment by bruvzg** [comment]
Some tests:

<details>
  <summary>Raw data</summary>

Average of 10 runs.

Container | Nr of iterations | Insert (µs) | Lookup (µs) | Iterate (µs) | Erase (µs)
-- | -- | -- | -- | -- | --
HashSet | 10000 | 178 | 14 | 15 | 21
HashSet | 50000 | 848 | 97 | 73 | 141
HashSet | 250000 | 4519 | 579 | 455 | 862
HashSet | 1000000 | 21052 | 3663 | 1790 | 6688
HashSet | 5000000 | 200373 | 36725 | 9166 | 54732
RBSet | 10000 | 713 | 386 | 17 | 397
RBSet | 50000 | 4266 | 2042 | 89 | 2114
RBSet | 250000 | 24769 | 11218 | 527 | 12005
RBSet | 1000000 | 121293 | 47999 | 1914 | 51400
RBSet | 5000000 | 723722 | 274974 | 10721 | 294013
std::set | 10000 | 525 | 356 | 33 | 737
std::set | 50000 | 2858 | 1886 | 170 | 3554
std::set | 250000 | 18098 | 10522 | 940 | 20897
std::set | 1000000 | 79986 | 45284 | 3998 | 90211
std::set | 5000000 | 422946 | 236493 | 18821 | 469186
unordered_set | 10000 | 240 | 21 | 13 | 196
unordered_set | 50000 | 899 | 112 | 66 | 1018
unordered_set | 250000 | 5385 | 607 | 336 | 5483
unordered_set | 1000000 | 34054 | 8335 | 1414 | 42392
unordered_set | 5000000 | 284808 | 78297 | 6756 | 356546

</details>

Screenshot 2022-05-19 at 18 53 06

**Comment by reduz** [comment]
@lawnjelly props seems your algorithm idea was pretty good!

@bruvzg what is the weirdest here is that iteration is faster in the STL one.  in HashSet iteration is literally just an array!

**Comment by mrjustaguy** [comment]
Eh, the graphs aren't marked.. what is time and what is the number of executions?
I'm guessing time is the y axis and the x axis the number of executions. is that right?

**Comment by bruvzg** [comment]
> I'm guessing time (ms?) is the y axis and the x axis the number of executions. is that right?

Y - time (µs, log scale), X - number of iterations.

**Comment by bruvzg** [comment]
> what is the weirdest here is that iteration is faster in the STL one. in HashSet iteration is literally just an array!

LLVM black magic? When it's built with GCC it's not faster.

The previous test was with LLVM 13, this one with GCC 11.3:

Screenshot 2022-05-20 at 12 16 42

**Comment by lawnjelly** [comment]
> @lawnjelly props seems your algorithm idea was pretty good!

As always, I would like to take credit when it goes right, and when it goes wrong it was all your idea. :grinning: 

I would actually be tempted to go further and try fully decoupling the two resizes (the keys versus the actual hashmap list). The keys could be stored as a vector and resized with a x2 basis (which reduces the resizes of the key, which could be expensive if they are big).

Also it might be an idea to try templating the `uint32_t` size for the IDs, because if you can guarantee the size will be less than 65535 you can use `uint16_t` which *may* be faster (may not of course, have to benchmark :grin: ).

As I say I think this approach of separating the expensive elements and using an ID in the hashmap list may also work well for the unordered hashmap.

File: core/templates/hash_set.h [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+#include "core/math/math_funcs.h"
+#include "core/os/memory.h"
+#include "core/templates/hash_map.h"
+#include "core/templates/hashfuncs.h"
+#include "core/templates/paged_allocator.h"
+
+/**
+ * Implementation of Set using a bidi indexed hash map.
+ * Use RBSet isntead of this only if the following conditions are met:
+ *
```
Comment (lawnjelly) [inline core/templates/hash_set.h:43]: "instead"

**Comment by akien-mga** [comment]
Leak sanitizer is detecting lots of leaks (in ShaderLanguage?), see https://github.com/godotengine/godot/runs/6528386338?check_suite_focus=true

**Review by akien-mga — APPROVED** [review]
Looks good to me, early benchmarks seem promising. I did a quick sanity check that editor startup and rendering of a simple 3D scene didn't regress (same speed as before).

**Comment by akien-mga** [comment]
Thanks!

**Comment by Calinou** [comment]
For reference, here's a benchmark of editor startup/shutdown on an empty project:

```
❯ hyperfine -m20 -iw1 "bin/godot.linuxbsd.tools.64.llvm.old /tmp/x/project.godot --quit" "bin/godot.linuxbsd.tools.64.llvm /tmp/x/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.tools.64.llvm.old /tmp/x/project.godot --quit
  Time (mean ± σ):      8.412 s ±  0.298 s    [User: 7.112 s, System: 0.358 s]
  Range (min … max):    7.533 s …  8.627 s    20 runs
 
Benchmark #2: bin/godot.linuxbsd.tools.64.llvm /tmp/x/project.godot --quit
  Time (mean ± σ):      8.337 s ±  0.276 s    [User: 7.100 s, System: 0.354 s]
  Range (min … max):    7.929 s …  8.600 s    20 runs
 
Summary
  'bin/godot.linuxbsd.tools.64.llvm /tmp/x/project.godot --quit' ran
    1.01 ± 0.05 times faster than 'bin/godot.linuxbsd.tools.64.llvm.old /tmp/x/project.godot --quit'
```

**Comment by reduz** [comment]
@Calinou all these changes show that, despite criticism that we were not using the most efficient data structures in most cases, in the end it makes little to no difference because these structures were not used for critical areas.

**Pull request merged**